File: Recursion/0002.Add Two Numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return dfs(l1, l2, false);
    }

    public ListNode dfs(ListNode l1, ListNode l2, boolean carry) {
        if (l1 == null && l2 == null) {
            return carry ? new ListNode(1) : null;
        }
        if (l1 == null) {
            l1 = l2;
            l2 = null;
        }
        int v = carry ? 1 : 0;
        v += l1.val;
        v += l2 == null ? 0 : l2.val;
        l1.val = v % 10;
        l1.next = dfs(l1.next, l2 == null ? null : l2.next, v >= 10);
        return l1;
    }
}

File: Recursion/0002.Add Two Numbers/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode], carry=False) -> Optional[ListNode]:
        if not l1 and not l2:
            return ListNode(1) if carry else None
        if not l1:
            l1, l2 = l2, l1
        v = 1 if carry else 0
        v += l1.val
        v += l2.val if l2 else 0
        l1.val = v % 10
        l1.next = self.addTwoNumbers(l1.next, l2.next if l2 else None, v >= 10)
        return l1